### ISSUE TITLE
OS X build failures

### ISSUE THREAD

**Issue by John-Colvin** [esa/pagmo#90]
After doing this: https://github.com/esa/pagmo/pull/89

I was able get the following steps to work:
```
$ cmake ../ -DBUILD_PYGMO=1
-- The C compiler identification is AppleClang 6.0.0.6000054
-- The CXX compiler identification is AppleClang 6.0.0.6000054
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- OS detected: Darwin
-- CXX Compiler detected: Clang
-- CMake Library search path: /usr/local/lib
-- Performing Test CLANG_TEMPLATE_DEPTH
-- Performing Test CLANG_TEMPLATE_DEPTH - Success
-- Enabling '-ftemplate-depth=256' compiler flag required since boost 1.54.
-- Performing Test ALL_C11
-- Performing Test ALL_C11 - Success
-- Enabling '-std=c++11' compiler flag
-- m library found: /usr/lib/libm.dylib
-- Found PythonInterp: /usr/bin/python (found version "2.7.6") 
-- Python version 2.7.6 found.
-- Found PythonInterp: /usr/bin/python2.7 (found suitable version "2.7.6", minimum required is "2.7.6") 
-- Found PythonLibs: /usr/lib/libpython2.7.dylib (found suitable version "2.7.6", minimum required is "2.7.6") 
-- Found PythonLibs: /usr/lib/libpython2.7.dylib (found version "2.7.6") 
-- Python interpreter: /usr/bin/python2.7
-- Python library: /usr/lib/libpython2.7.dylib
-- Python include path: /System/Library/Frameworks/Python.framework/Headers
-- Python packages dir is: site-packages
-- OSX system detected.
-- Python interpeter returns string: 'Python 2.7.6'
-- Python library version: 27
-- Python modules install path: lib/python2.7/site-packages
-- Required Boost libraries: system;serialization;thread;python
-- Boost version: 1.56.0
-- Found the following Boost libraries:
--   system
--   serialization
--   thread
--   python
-- Detected Boost version: 105600
-- Boost include dirs: /usr/local/include
-- Boost libraries: /usr/local/lib/libboost_system-mt.dylib;/usr/local/lib/libboost_serialization-mt.dylib;/usr/local/lib/libboost_thread-mt.dylib;/usr/local/lib/libboost_python-mt.dylib
-- Looking for include file pthread.h
-- Looking for include file pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE  
-- Thread library: 
-- Build flags:  -ftemplate-depth=256 -std=c++11 -DBOOST_THREAD_USE_DLL -DBOOST_SERIALIZATION_DYN_LINK=1
-- Module linker flags: 
-- Shared linker flags: 
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/john/Git/pagmo/build
```
followed by
```
$ cmake ../ -DBUILD_PYGMO=1 -DPYGMO_PYTHON_VERSION=3.4.2 -DCMAKE_PREFIX_PATH=/usr/local/Frameworks/Python.framework/Versions/3.4/
-- OS detected: Darwin
-- CXX Compiler detected: Clang
-- CMake Library search path: /usr/local/lib
-- Enabling '-ftemplate-depth=256' compiler flag required since boost 1.54.
-- Enabling '-std=c++11' compiler flag
-- m library found: /usr/lib/libm.dylib
-- Found PythonInterp: /usr/local/Frameworks/Python.framework/Versions/3.4/bin/python3.4 (found suitable version "3.4.2", minimum required is "3.4.2") 
-- Could NOT find PythonLibs: Found unsuitable version "2.7.6", but required is at least "3.4.2" (found /usr/local/Frameworks/Python.framework/Versions/3.4/lib/libpython3.4m.dylib)
-- Found PythonLibs: /usr/local/Frameworks/Python.framework/Versions/3.4/lib/libpython3.4m.dylib (found version "2.7.6") 
-- Python interpreter: /usr/local/Frameworks/Python.framework/Versions/3.4/bin/python3.4
-- Python library: /usr/local/Frameworks/Python.framework/Versions/3.4/lib/libpython3.4m.dylib
-- Python include path: /System/Library/Frameworks/Python.framework/Headers
-- Python packages dir is: site-packages
-- OSX system detected.
-- Python interpeter returns string: 'Python 3.4.2'
-- Python library version: 34
-- Python modules install path: lib/python3.4/site-packages
-- Required Boost libraries: system;serialization;thread;python3
-- Boost version: 1.56.0
-- Found the following Boost libraries:
--   system
--   serialization
--   thread
--   python3
-- Detected Boost version: 105600
-- Boost include dirs: /usr/local/include
-- Boost libraries: /usr/local/lib/libboost_system-mt.dylib;/usr/local/lib/libboost_serialization-mt.dylib;/usr/local/lib/libboost_thread-mt.dylib;/usr/local/lib/libboost_python3-mt.dylib
-- Thread library: 
-- Build flags:  -ftemplate-depth=256 -std=c++11 -DBOOST_THREAD_USE_DLL -DBOOST_SERIALIZATION_DYN_LINK=1
-- Module linker flags: 
-- Shared linker flags: 
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/john/Git/pagmo/build
```

so aside from the weird stuff going on with PythonLibs things are looking ok.

I then run:
```$ make```
and everything compiles happily, but the linking step fails:
```
Linking CXX shared library _algorithm.so
Undefined symbols for architecture x86_64:
  "_PyInt_FromLong", referenced from:
      boost::python::api::proxy<boost::python::api::item_policies> boost::python::api::object_operators<boost::python::api::object>::operator[]<int>(int const&) in algorithm.cpp.o
      boost::python::objects::caller_py_function_impl<boost::python::detail::caller<int (pagmo::algorithm::cmaes::*)() const, boost::python::default_call_policies, boost::mpl::vector2<int, pagmo::algorithm::cmaes&> > >::operator()(_object*, _object*) in algorithm.cpp.o
  "_PyInt_Type", referenced from:
      boost::python::detail::converter_target_type<boost::python::to_python_value<int const&> >::get_pytype() in algorithm.cpp.o
  "_PyString_FromStringAndSize", referenced from:
      boost::python::detail::caller_arity<1u>::impl<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > (*)(pagmo::algorithm::sa_corana const&), boost::python::default_call_policies, boost::mpl::vector2<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, pagmo::algorithm::sa_corana const&> >::operator()(_object*, _object*) in algorithm.cpp.o
      boost::python::tuple boost::python::make_tuple<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in algorithm.cpp.o
      boost::python::detail::caller_arity<1u>::impl<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > (*)(pagmo::algorithm::de_1220 const&), boost::python::default_call_policies, boost::mpl::vector2<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, pagmo::algorithm::de_1220 const&> >::operator()(_object*, _object*) in algorithm.cpp.o
      boost::python::detail::caller_arity<1u>::impl<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > (*)(pagmo::algorithm::mde_pbx const&), boost::python::default_call_policies, boost::mpl::vector2<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, pagmo::algorithm::mde_pbx const&> >::operator()(_object*, _object*) in algorithm.cpp.o
      boost::python::detail::caller_arity<1u>::impl<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > (*)(pagmo::algorithm::jde const&), boost::python::default_call_policies, boost::mpl::vector2<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, pagmo::algorithm::jde const&> >::operator()(_object*, _object*) in algorithm.cpp.o
      boost::python::detail::caller_arity<1u>::impl<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > (*)(pagmo::algorithm::de const&), boost::python::default_call_policies, boost::mpl::vector2<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, pagmo::algorithm::de const&> >::operator()(_object*, _object*) in algorithm.cpp.o
      boost::python::detail::caller_arity<1u>::impl<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > (*)(pagmo::algorithm::spea2 const&), boost::python::default_call_policies, boost::mpl::vector2<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, pagmo::algorithm::spea2 const&> >::operator()(_object*, _object*) in algorithm.cpp.o
      ...
  "_PyString_Type", referenced from:
      boost::python::detail::converter_target_type<boost::python::to_python_value<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&> >::get_pytype() in algorithm.cpp.o
  "boost::python::detail::init_module(char const*, void (*)())", referenced from:
      _init_algorithm in algorithm.cpp.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [PyGMO/algorithm/_algorithm.so] Error 1
make[1]: *** [PyGMO/algorithm/CMakeFiles/_algorithm.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by darioizzo** [comment]
I will close this as it looks like its the user that is mixing pythons versions in OSX. With homebrew python and boost (2.7) pagmo has been tested and build succeed.

**Comment by John-Colvin** [comment]
I have homebrew python 3 and boost-python (with python3 support) installed, of course alongside the default os x Python 2. I haven't had any problems with other libraries relating to python.

**Comment by darioizzo** [comment]
Thats weird, I also have homebrew and python 2.7 but all works for me. 
I have compiled boost and python using homebrew and use these versions. If you use the systemwide python (from OSX) PyGMO will not work. 

I have never tried on osx python 3 and PyGMO, on linux this works just fine, but, as I said, I have tried on osx, I might give it a try to see if I can help you there .... In case I can reproduce your problem I will reopen the ticket and try to solve ...
